### PR TITLE
python3Packages.tensorflow-datasets: init at 4.4.0

### PR DESCRIPTION
**Pull request by ndl**
###### Motivation for this change

`tensorflow-datasets` is a collection of ML datasets with a common API to access these that's often used both in TF-based and other ML applications.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/tensorflow-datasets/default.nix [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    "tensorflow_datasets/scripts/documentation/build_api_docs_test.py"
+
+    # Not a test, should not be executed.
+    "tensorflow_datasets/testing/test_utils.py"
+
+    # Require `gcld3` and `nltk.punkt` which are not packaged in `nixpkgs`.
+    "tensorflow_datasets/text/c4_test.py"
+    "tensorflow_datasets/text/c4_utils_test.py"
+
+    # Requires `bs4` which is not packaged in `nixpkgs`.
```
Comment (veprbl) [inline pkgs/development/python-modules/tensorflow-datasets/default.nix:131]: python3Packages.beautifulsoup4

Comment (ndl) [reply pkgs/development/python-modules/tensorflow-datasets/default.nix]: Oh, missed that one - thanks! The test now passes => removed it from the exclusions.

File: pkgs/development/python-modules/tensorflow-datasets/corruptions.patch [hunk 2]
```diff
@@ -0,0 +1,22 @@
+diff --git a/tensorflow_datasets/image_classification/corruptions.py b/tensorflow_datasets/image_classification/corruptions.py
+index 066c4460..cb9a6667 100644
+--- a/tensorflow_datasets/image_classification/corruptions.py
++++ b/tensorflow_datasets/image_classification/corruptions.py
+@@ -35,7 +35,7 @@ FROST_FILENAMES = []
+ 
+ 
+ def _imagemagick_bin():
+-  return 'imagemagick'  # pylint: disable=unreachable
++  return 'convert'  # pylint: disable=unreachable
```
Comment (samuela) [inline pkgs/development/python-modules/tensorflow-datasets/corruptions.patch:10]: Why is our binary name for imagemagick different?

File: pkgs/development/python-modules/tensorflow-datasets/corruptions.patch [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ 
+ 
+ # /////////////// Corruption Helpers ///////////////
+@@ -675,7 +675,7 @@ def spatter(x, severity=1):
+     #     ker = np.array([[-1,-2,-3],[-2,0,0],[-3,0,1]], dtype=np.float32)
+     #     ker -= np.mean(ker)
+     ker = np.array([[-2, -1, 0], [-1, 1, 1], [0, 1, 2]])
+-    dist = cv2.filter2D(dist, cv2.CVX_8U, ker)
++    dist = cv2.filter2D(dist, cv2.CV_8U, ker)
```
Comment (samuela) [inline pkgs/development/python-modules/tensorflow-datasets/corruptions.patch:19]: Why this change?

File: pkgs/development/python-modules/tensorflow-datasets/default.nix [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    "tensorflow_datasets/core/utils/gcs_utils_test.py"
+    "tensorflow_datasets/scripts/cli/build_test.py"
+
+    # Requires `pretty_midi` which is not packaged in `nixpkgs`.
+    "tensorflow_datasets/audio/groove_test.py"
+
+    # Requires `crepe` which is not packaged in `nixpkgs`.
+    "tensorflow_datasets/audio/nsynth_test.py"
+
+    # Requires multiple packages missing in `nixpkgs`.
```
Comment (samuela) [inline pkgs/development/python-modules/tensorflow-datasets/default.nix:112]: Which ones?

File: pkgs/development/python-modules/tensorflow-datasets/default.nix [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    # to the differences in some of the dependencies?
+    "tensorflow_datasets/rl_unplugged/rlu_atari/rlu_atari_test.py"
+
+    # Requires `tensorflow_docs` which is not packaged in `nixpkgs` and the test is for documentation anyway.
+    "tensorflow_datasets/scripts/documentation/build_api_docs_test.py"
+
+    # Not a test, should not be executed.
+    "tensorflow_datasets/testing/test_utils.py"
+
+    # Require `gcld3` and `nltk.punkt` which are not packaged in `nixpkgs`.
```
Comment (samuela) [inline pkgs/development/python-modules/tensorflow-datasets/default.nix:129]: I can't find gcld3 but isn't [this nltk](https://search.nixos.org/packages?channel=21.11&from=0&size=50&buckets=%7B%22package_attr_set%22%3A%5B%22python39Packages%22%5D%2C%22package_license_set%22%3A%5B%5D%2C%22package_maintainers_set%22%3A%5B%5D%2C%22package_platforms%22%3A%5B%5D%7D&sort=relevance&type=packages&query=nltk)? Is there something about the punkt submodule that doesn't work?

**Review by samuela — CHANGES_REQUESTED** [review]
Just a few questions/clarifications

File: pkgs/development/python-modules/tensorflow-datasets/corruptions.patch (hunk 2 above)
Comment (ndl) [reply pkgs/development/python-modules/tensorflow-datasets/corruptions.patch:10]: All distributions I've checked (NixOS, Gentoo, Ubuntu) provide `convert` binary for ImageMagick 6.x and `magick` for ImageMagick 7.x (but with fallback link for `convert`). I have no idea where they've got `imagemagick` binary name from. Based on other evidence (see below) it's not clear whether this test worked any time recently?

File: pkgs/development/python-modules/tensorflow-datasets/corruptions.patch (hunk 3 above)
Comment (ndl) [reply pkgs/development/python-modules/tensorflow-datasets/corruptions.patch:19]: I don't believe there's `CVX_8U` data type in OpenCV, this looks like a typo. The test fails without this patch due to `CVX_8U` not found but succeeds once fixed as above.

File: pkgs/development/python-modules/tensorflow-datasets/default.nix (hunk 4 above)
Comment (ndl) [reply pkgs/development/python-modules/tensorflow-datasets/default.nix]: Actually, after Dmitry noticed above that we do in fact have `bs4`, there are just two - changed the comment.

File: pkgs/development/python-modules/tensorflow-datasets/default.nix (hunk 5 above)
Comment (ndl) [reply pkgs/development/python-modules/tensorflow-datasets/default.nix:129]: NLTK doesn't bundle the data into its package - it downloads the data "on demand", which doesn't work in sandbox. See also https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/nltk/default.nix#L25 I think ideally NLTK should be packaged in a modular way so that its datasets are provided as independent packages, but I don't know how difficult that's going to be and that's way beyond the scope of this PR anyway :-)

File: pkgs/development/python-modules/tensorflow-datasets/corruptions.patch (hunk 2 above)
Comment (samuela) [reply pkgs/development/python-modules/tensorflow-datasets/corruptions.patch:10]: Ah ok, got it! Seems like it's also flagged as unreachable code by pylint anyhow.

File: pkgs/development/python-modules/tensorflow-datasets/default.nix (hunk 5 above)
Comment (samuela) [reply pkgs/development/python-modules/tensorflow-datasets/default.nix:129]: ah ok makes sense!

File: pkgs/development/python-modules/tensorflow-datasets/corruptions.patch (hunk 3 above)
Comment (samuela) [reply pkgs/development/python-modules/tensorflow-datasets/corruptions.patch:19]: hmm weird is there an upstream issue for this?

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 154117` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>python310Packages.tensorflow-datasets</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>python39Packages.tensorflow-datasets</li>
  </ul>
</details>

File: pkgs/development/python-modules/tensorflow-datasets/corruptions.patch (hunk 3 above)
Comment (ndl) [reply pkgs/development/python-modules/tensorflow-datasets/corruptions.patch:19]: Didn't find one, created https://github.com/tensorflow/datasets/issues/3673

**Review by samuela — APPROVED** [review]
LGTM I'll merge tomorrow unless anyone objects

File: pkgs/development/python-modules/tensorflow-datasets/default.nix [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+  src = fetchFromGitHub {
+    owner = "tensorflow";
+    repo = "datasets";
+    rev = "v${version}";
+    sha256 = "11kbpv54nwr0xf7z5mkj2lmrfqfmcdq8qcpapnqck1kiawr3yad6";
+  };
+
+  patches = [
+    ./corruptions.patch
```
Comment (veprbl) [inline pkgs/development/python-modules/tensorflow-datasets/default.nix:51]:
```suggestion
    # addresses https://github.com/tensorflow/datasets/issues/3673
    ./corruptions.patch
```

**Comment by ndl** [comment]
@samuela thanks for all the help and timely reviews on all these PRs, really appreciated! :-)

**Comment by samuela** [comment]
Absolutely, thank you for the PRs @ndl! I will def use these packages myself!
